# Parameter tunning

### proxy strength (more or less informative about U): sigma_z, sigma_v

| Corr(Z, U) | σ (sigma) |
|------------|-----------|
| 0.9        | 0.73      |
| 0.8        | 1.125     |
| 0.7        | 1.53      |
| 0.6        | 2.0       |
| 0.5        | 2.6       |
| 0.4        | 3.44      |
| 0.3        | 4.77      |
| 0.2        | 7.35      |
| 0.1        | 14.93     |

* we fix $a_z, a_v = 1.5$ for all cases.
* The table for sigma_v are the same as sigma_z.

### strength of confounding (how much U affects treatment assignment W and outcome T): gamma_u_in_a, beta_u_in_t


| level    | gamma_u_in_a | beta_u_in_t |
|----------|---------|--------|
| none     | 0.0     | 0.0    |
| weak     | 0.2     | 0.2    |
| moderate | 0.5     | 0.5    |
| strong   | 1.0     | 1.0    |
| extreme  | 2.0     | 2.0    |

### strength of treatment effect: tau_log_hr

| tau_log_hr value | Interpretation                          |
|------------------|------------------------------------------|
| 0                | No treatment effect                      |
| < 0              | Beneficial (lower hazard)      |
| > 0              | Harmful (higher hazard)        |

**possible range**

| Effect size category        | tau_log_hr range | Hazard ratio range (exp(tau)) | Interpretation |
|----------------------------|------------------|-------------------------------|----------------|
| Null / negligible          | [-0.1, 0.1]      | [0.90, 1.11]                  | Little to no effect |
| Small to moderate (typical)| [-0.3, 0.3]      | [0.74, 1.35]                  | Plausible clinical effects |
| Moderate to large          | [-0.7, 0.7]      | [0.50, 2.01]                  | Strong but still realistic |
| Extreme (use with caution) | < -1 or > 1      | < 0.37 or > 2.7               | Often unrealistic / unstable |

# Testing

In [1]:
# Add parent directory to Python path to import nc_csf module
import sys
from pathlib import Path

# Get the parent directory (where nc_csf folder is located)
parent_dir = Path.cwd().parent
if str(parent_dir) not in sys.path:
    sys.path.insert(0, str(parent_dir))
    
print(f"Added to path: {parent_dir}")

Added to path: c:\Users\17414\OneDrive\桌面\USC\negavitve-control-proxies-causual-survival-forest


In [2]:
from nc_csf.evaluate_performance import run_experiment

## Linear in treatment and outcome

### Informative proxies & strong confounding & harmful large treatment effect

In [3]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.9,
    k_t=1.5, lam_t=0.4, tau_log_hr=-0.7,beta_u_in_t=1.1,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=1.125, aW=1.5, sigma_w=1.53,
    linear_treatment=True, linear_outcome=True
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE     Slope
Baseline 0.499859 0.344711 -0.344422  -0.164877   -0.110721  1.426097 -0.023747
  NC-CSF 0.380873 0.206878 -0.146033   0.042803    0.138897  1.086631  0.005901
  Oracle 0.334110 0.210038  0.000057   0.327956    0.426891  0.953216  0.065889


### Informative proxies & weak confounding & benificial small treatment effect

In [4]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.2,
    k_t=1.5, lam_t=0.4, tau_log_hr=0.25,beta_u_in_t=0.2,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=1.125, aW=1.5, sigma_w=1.53,
    linear_treatment=True, linear_outcome=True
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE    Slope
Baseline 0.040801 0.032616 -0.026617   0.744210    0.726958  0.947402 0.750467
  NC-CSF 0.034339 0.026291 -0.010252   0.675835    0.606953  0.797343 0.585082
  Oracle 0.033416 0.025420 -0.006119   0.658328    0.582392  0.775907 0.515004


### Non informative proxies & strong confounding & benificial moderate treatment effect

In [5]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.9,
    k_t=1.5, lam_t=0.4, tau_log_hr=0.5,beta_u_in_t=1.1,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=7.35, aW=1.5, sigma_w=4.77,
    linear_treatment=True, linear_outcome=True
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE    Slope
Baseline 0.278350 0.250874 -0.228615   0.432635    0.515404  1.665955 0.317446
  NC-CSF 0.163766 0.119047 -0.051746   0.385846    0.463015  0.980159 0.194123
  Oracle 0.155583 0.092865  0.010310   0.370863    0.439841  0.931183 0.126919


### Non informative proxies & weak confounding & near null treatment effect

In [6]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.25,
    k_t=1.5, lam_t=0.4, tau_log_hr=0.12,beta_u_in_t=0.2,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=7.35, aW=1.5, sigma_w=4.77,
    linear_treatment=True, linear_outcome=True
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE    Slope
Baseline 0.038044 0.028464 -0.026188   0.676184    0.608051  1.763885 1.165317
  NC-CSF 0.023403 0.017719 -0.003434   0.537232    0.362932  1.085081 0.645680
  Oracle 0.025310 0.018825  0.002459   0.432010    0.213929  1.173481 0.507561


## Non linear in only outcome

### Informative proxies & strong confounding & harmful large treatment effect

In [7]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.9,
    k_t=1.5, lam_t=0.4, tau_log_hr=-0.7,beta_u_in_t=1.1,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=1.125, aW=1.5, sigma_w=1.53,
    linear_treatment=True, linear_outcome=False
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE     Slope
Baseline 0.494029 0.339675 -0.339632  -0.146138   -0.101169  1.417188 -0.020135
  NC-CSF 0.378413 0.201295 -0.150206   0.092115    0.147821  1.085528  0.011673
  Oracle 0.333253 0.203394 -0.008754   0.329233    0.423019  0.955979  0.059980


### Informative proxies & weak confounding & benificial small treatment effect

In [8]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.2,
    k_t=1.5, lam_t=0.4, tau_log_hr=0.3,beta_u_in_t=0.2,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=1.125, aW=1.5, sigma_w=1.53,
    linear_treatment=True, linear_outcome=False
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE    Slope
Baseline 0.042552 0.035026 -0.026473   0.741166    0.742543  0.862856 0.605623
  NC-CSF 0.039254 0.029079 -0.008205   0.631993    0.600415  0.795977 0.445780
  Oracle 0.038911 0.028695 -0.003631   0.618794    0.557570  0.789031 0.387715


### Non informative proxies & strong confounding & benificial moderate treatment effect

In [9]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.9,
    k_t=1.5, lam_t=0.4, tau_log_hr=0.5,beta_u_in_t=1.1,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=7.35, aW=1.5, sigma_w=4.77,
    linear_treatment=True, linear_outcome=False
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE    Slope
Baseline 0.277472 0.252680 -0.228395   0.399473    0.489287  1.669800 0.256515
  NC-CSF 0.162705 0.117605 -0.050555   0.374345    0.451260  0.979146 0.169968
  Oracle 0.156869 0.092516  0.010566   0.338759    0.390083  0.944027 0.100056


### Non informative proxies & weak confounding & near null treatment effect

In [10]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.25,
    k_t=1.5, lam_t=0.4, tau_log_hr=0.12,beta_u_in_t=0.2,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=7.35, aW=1.5, sigma_w=4.77,
    linear_treatment=True, linear_outcome=False
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE    Slope
Baseline 0.040837 0.030379 -0.028366   0.630907    0.553925  1.952400 1.137445
  NC-CSF 0.027543 0.019985 -0.001976   0.369052    0.191091  1.316823 0.479021
  Oracle 0.025573 0.018364  0.004086   0.331799    0.212079  1.222648 0.360128


## Non linear in treatment & outcome

### Informative proxies & strong confounding & harmful large treatment effect

In [11]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.9,
    k_t=1.5, lam_t=0.4, tau_log_hr=-0.7,beta_u_in_t=1.1,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=1.125, aW=1.5, sigma_w=1.53,
    linear_treatment=False, linear_outcome=False
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE    Slope
Baseline 0.396208 0.187211 -0.084108   0.338777    0.403050  0.967129 0.084606
  NC-CSF 0.387055 0.201557 -0.021496   0.332914    0.392468  0.944788 0.119743
  Oracle 0.385780 0.203217 -0.003553   0.349352    0.428132  0.941674 0.089479


### Informative proxies & weak confounding & benificial small treatment effect

In [12]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.2,
    k_t=1.5, lam_t=0.4, tau_log_hr=0.3,beta_u_in_t=0.2,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=1.125, aW=1.5, sigma_w=1.53,
    linear_treatment=False, linear_outcome=False
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE    Slope
Baseline 0.040068 0.031516 -0.020090   0.693845    0.696322  0.842740 0.561138
  NC-CSF 0.039175 0.030371 -0.017618   0.683969    0.697805  0.823963 0.534567
  Oracle 0.037478 0.028888 -0.015431   0.698287    0.701219  0.788256 0.529910


### Non informative proxies & strong confounding & benificial moderate treatment effect

In [13]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.9,
    k_t=1.5, lam_t=0.4, tau_log_hr=0.5,beta_u_in_t=1.1,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=7.35, aW=1.5, sigma_w=4.77,
    linear_treatment=False, linear_outcome=False
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE    Slope
Baseline 0.192856 0.120944 -0.040552   0.267430    0.397330  0.987567 0.055321
  NC-CSF 0.193037 0.101850  0.006112   0.159279    0.262289  0.988488 0.031559
  Oracle 0.189869 0.096817  0.012917   0.249050    0.356204  0.972267 0.048522


### Non informative proxies & weak confounding & near null treatment effect

In [14]:
results_df, cate_te, pred_baseline, pred_nccsf, pred_oracle = run_experiment(
    n=5000, p_x=10, seed=123,
    a_prevalence=0.5, gamma_u_in_a=0.25,
    k_t=1.5, lam_t=0.4, tau_log_hr=0.12,beta_u_in_t=0.2,
    k_c=1.2, lam_c=1e6, beta_u_in_c=0.3, target_censor_rate=0, max_censor_calib_iter=60, censor_lam_lo=1e-8,censor_lam_hi=1e6, admin_censor_time=None,
    aZ=1.5, sigma_z=7.35, aW=1.5, sigma_w=4.77,
    linear_treatment=False, linear_outcome=False
)

print("\nEvaluation Results:")
print(results_df.to_string(index=False))


Evaluation Results:
   Model     RMSE      MAE      Bias  Pearson r  Spearman r  Rel RMSE    Slope
Baseline 0.031957 0.025620 -0.022980   0.648559    0.614484  1.584702 0.937400
  NC-CSF 0.029899 0.023181 -0.019633   0.612414    0.602950  1.482669 0.859973
  Oracle 0.029294 0.022683 -0.019323   0.648224    0.632218  1.452642 0.928002


---

# Summary of Results (RMSE)

## Linear Treatment & Outcome

| Model | Informative + Strong | Informative + Weak | Non-informative + Strong | Non-informative + Weak |
|-------|---------------------|--------------------|--------------------------|-----------------------|
| Baseline | 0.500 | 0.041 | 0.278 | 0.038 |
| NC-CSF | 0.381 | 0.034 | 0.164 | 0.023 |
| Oracle | 0.334 | 0.033 | 0.156 | 0.025 |

## Nonlinear Outcome Only

| Model | Informative + Strong | Informative + Weak | Non-informative + Strong | Non-informative + Weak |
|-------|---------------------|--------------------|--------------------------|-----------------------|
| Baseline | 0.494 | 0.043 | 0.277 | 0.041 |
| NC-CSF | 0.378 | 0.039 | 0.163 | 0.028 |
| Oracle | 0.333 | 0.039 | 0.157 | 0.026 |

## Nonlinear Treatment & Outcome

| Model | Informative + Strong | Informative + Weak | Non-informative + Strong | Non-informative + Weak |
|-------|---------------------|--------------------|--------------------------|-----------------------|
| Baseline | 0.396 | 0.040 | 0.193 | 0.032 |
| NC-CSF | 0.387 | 0.039 | 0.193 | 0.030 |
| Oracle | 0.386 | 0.037 | 0.190 | 0.029 |
